torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -751,24 +758,47 @@ void Reducer::initialize_buckets(
 // (see Note:  "Gradient Layout Contract" in initialize_buckets).
 void Reducer::initialize_bucketviews(
     Reducer::BucketReplica& replica,
-    at::Tensor& contents) {
+    at::Tensor& contents,
+    bool copy_to_bucket_view) {
   for (size_t i = 0; i < replica.variables.size(); i++) {
-    const auto& v = replica.variables[i];
+    auto& v = replica.variables[i];
     const auto offset = replica.offsets[i];
     const auto length = replica.lengths[i];
+    at::Tensor bucket_view;
     if (v.is_non_overlapping_and_dense()) {
       // If the param's memory is dense, match its layout, anticipating
       // the autograd engine (AccumulateGrad) will also create gradients
       // matching its layout.
-      replica.bucket_views.push_back(
-          contents.as_strided(v.sizes(), v.strides(), offset));
+      bucket_view = contents.as_strided(v.sizes(), v.strides(), offset);
     } else {
       // Fall back to a C-style contiguous view, again anticipating
       // AccumulateGrad will do the same when stashing grads for non-dense
       // params.
-      replica.bucket_views.push_back(
-          contents.narrow(0, offset, length).view(v.sizes()));
+      bucket_view = contents.narrow(0, offset, length).view(v.sizes());
     }
+    replica.bucket_views.push_back(bucket_view);
+    // If grad has already been defined/calculated in previous iterations,
+    // and the grad does not point to bucket_view, that means bucket_view
+    // is newly allocated, if this function is called inside initialize_bucket,
+    // old grad needs to be copied into new bucket_view for initialization
+    // and let grad point to bucket_view;
+    // otherwise this function is called for communication hook, bucket_view
+    // has the updated results in new tensor, just let grad point to
+    // bucket_view. If grad is not defined, do not let it point to bucket_view.
+    // E.g., for global_unused parameters, their grads should be kept as being
+    // undefined.
+    runGradCallbackForVariable(v, [&](auto& grad) {
+      if (grad.defined() && !grad.is_alias_of(bucket_view)) {
```

Comment:
I'm wondering if my understanding is correct about what happens in the first iteration:

1. In the first iteration, the grads are not defined when we enter here via `initialize_buckets`. 
2. As a result, the grad has separate storage and the bucket has separate storage.
3. When we execute the logic in `mark_variable_ready_dense`, the grad produced by backward is copied into the bucket and now the grad points to the bucket instead of having its own storage.
4. Now, from the second iteration onward it seems like we'll have only one copy?

Does this mean we'll still have two copies of the gradient very briefly for each parameter? I guess this should be fine in most cases, but in some cases a parameter could be pretty large (large nn.EmbeddingBag). 

> Also be noted that we can not make variable.grad() pointing to bucket buffer during construction time, because we want to
keep grad undefined for unused parameters.

I was wondering if we can just set the gradients of the unused parameters as undefined at the end of the backward pass to achieve this? At the beginning of the next forward pass, we can make these grads point back to the bucket.

Reply:
for 4. second iteration will call initialize_bucket again to rebuild buckets. if users did not zero out grad or set grad to be none in their user codes, we need to copy old bucket to new bucket in this case, and maintain two copies for the bucket, when parameter is large, the memory saving is not large. I'm trying to add codes where I will check if grad is zero tensor, I will free the grad before allocating a new bucket. it will work well for most cases when users usually zero out grads. 

"I was wondering if we can just set the gradients of the unused parameters as undefined at the end of the backward pass to achieve this? At the beginning of the next forward pass, we can make these grads point back to the bucket."
==> this sounds like a good idea! possibly it can make codes cleaner, but how to set gradient to be undefined?

Reply:
Had second thought. I think we can not set gradients of the unused parameters as undefined at the end of the backward pass, because if the parameter is unused, we just want to keep the grad not updated, not keep the grad undefined.

Reply:
I'm wondering if we can just have a flag that indicates whether the parameter was undefined or not at the beginning  of the backward pass? The process would be something like this:

1. For the first iteration, if grads are not undefined, we copy them to the bucket and make the grads point to the bucket. For undefined grads, we just point them to a bucket with all zeros. For each parameter we store a flag of whether or not they were undefined at the beginning of the backward pass.
2. Once the backward pass is complete, if a parameter was undefined at the beginning of the backward pass and it was unused, we set its grad to None.
3. For the next backward pass, we repeat the same process where we identify the undefined parameters, point them back to the bucket and keep a flag if they were undefined at the beginning of the backward pass.

Reply:
sounds reasonable, I think your suggestion can help further avoid copy if grad is none before backward starts in some iterations. The number of grads usually is hundreds, so extra memory usage is trivial as well.

Reply:
Had thought about it again:). The above suggestion assumes grads will be mutated in backward pass in place.  ===> but it is not guaranteed.   

if grads are mutated out of place in backward, above suggestion will not give correct results. 

although current diff (let grad point to bucket_view when grad is ready) can not save memory when grads are mutated out of place, it can still give correct results.

How do you think? I feel it is more safe to let grad point to bucket_view during backward in this way

Reply:
Yes, you are right it won't work in the case where the user might modify the grads themselves. I think this PR looks good to me and I'll let @mrshenli stamp it.

Although, can we add some unit tests to verify everything works as expected with this PR when the user modifies `.grad` field themselves? These tests would ensure our logic for copying the grads in this case work as expected.

Reply:
sounds good, let me add one more test for this case

Reply:
actually I checked our unit tests that have already manually set grad to be none after iteration, this will trigger copying the grads here. and it is working as expected